Guide for webauthn authenticator in dotnet (#3021)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/authenticators-web-authn/main/aspnet/enrollmentintegrationsteps.md

### 1: Build a sign-in page on the client

> Build a sign in page that captures the user’s name and password, as shown in the following example.

Author: done

![Basic sign-in dialog](/img/authenticators/dotnet-authenticators-signinform.png)

### 2: Authenticate the user credentials

> Once a user has initiated the sign-on process by entering the username and password and clicking **Log In**, create an `AuthenticationOptions` object in your `LogIn` method and set its `Username` and `Password` properties to the values set by the user. Pass this object as a parameter to the `AuthenticateAsync` method on the `IdxClient` which you have also instantiated.

Author: @susanharper-okta I guess it isn't essential. It's more a question of whether or not you need to cover where the IdxClient comes from. I gues you can't hand hold the reader through everything.

```csharp
var _idxClient = new Okta.Idx.Sdk.IdxClient();
var authnOptions = new Okta.Idx.Sdk.AuthenticationOptions
{
    Username = model.UserName,
    Password = model.Password,
};

try
{
    var authnResponse = await _idxClient.AuthenticateAsync(authnOptions).ConfigureAwait(false);
```

### 3: Handle the response from the sign-in flow

Query the `AuthenticationStatus` property of the `AuthenticationResponse` object returned by `AuthenticateAsync` to discover the current status of the authentication process.

```csharp
    Session["idxContext"] = authnResponse.IdxContext;
    switch (authnResponse?.AuthenticationStatus)
    {
        case AuthenticationStatus.Success:
                … your code …
        case AuthenticationStatus.PasswordExpired:
                … your code …
```

> If you've configured your Okta org correctly, you'll need to respond to two specific authenticator statuses to handle this scenario in addition to `Success` and `PasswordExpired`:

Author: done

> * `AwaitingAuthenticatorEnrollment` which is covered in this section

Author: done

> * `AwaitingChallengeAuthenticatorSelection` which is covered in the [challenge flow section](/docs/guides/authenticators-web-authn/aspnet/main/#integrate-sdk-for-authenticator-challenge).

Author: done

> The names of the authenticators available for enrollment or challenge can be found in the `AuthenticationResponse` object's `Authenticators` collection. The user should be redirected to the authenticator list page that displays the WebAuthn Authenticator factor as an authenticator to be verified.

Author: done

> > Note that the `isChallengeFlow` session variable is set to `false` if the user has not already enrolled, and `true` if they have already done so.

Author: As previous comment. We're enrolling the authenticator at this point, not the user.

```csharp
public ActionResult SelectAuthenticator()
{
   var authenticators = 
      (IList<AuthenticatorViewModel>)Session["authenticators"] ?? 
      new List<AuthenticatorViewModel>();
   var viewModel = new SelectAuthenticatorViewModel
   {
      Authenticators = authenticators,
      AuthenticatorId = authenticators.
         FirstOrDefault()?.AuthenticatorId,
      PasswordId = authenticators.
         FirstOrDefault(x => x.Name.ToLower() == "password")?.AuthenticatorId,
      PhoneId = authenticators.
         FirstOrDefault(x => x.Name.ToLower() == "phone")?.AuthenticatorId,
      WebAuthnId = authenticators.
         FirstOrDefault(x => x.Name.ToLower() == "security key or biometric")?.AuthenticatorId,
      TotpId = authenticators.
         FirstOrDefault(x => x.Name.ToLower() == "google authenticator")?.AuthenticatorId,
      OktaVerifyId = authenticators.
         FirstOrDefault(x => x.Name.ToLower() == "okta verify")?.AuthenticatorId,
      CanSkip = TempData["canSkip"] != null && (bool)TempData["canSkip"]
   };
   return View(viewModel);
}
```

> This `viewModel` is then consumed in a Razor page.

Author: Done

       new { ReturnUrl = ViewBag.ReturnUrl }, FormMethod.Post,
       new { @class = "form-horizontal", role = "form" }))
    {
        @Html.AntiForgeryToken()
        <!-- Headings elided -->
        @Html.ValidationSummary(true, "", new { @class = "text-danger" })
        @Html.HiddenFor(m => m.PasswordId)
        @Html.HiddenFor(m => m.PhoneId)
        @Html.HiddenFor(m => m.WebAuthnId)
        @Html.HiddenFor(m => m.OktaVerifyId)
        @Html.HiddenFor(m => m.TotpId)
        <ul>
            @foreach (var authenticator in Model.Authenticators)
            {
                <div>
                    <label>
                        @Html.RadioButtonFor(m => m.AuthenticatorId,
                            authenticator.AuthenticatorId)
                        @authenticator.Name
                    </label>
                </div>
            }
        </ul>
        @Html.ValidationMessageFor(m => m.AuthenticatorId,
            "", new { @class = "text-danger" })
        <div class="form-group">
            <div>
                <input type="submit" value="Submit" />
            </div>
        </div>
    }
</section>
```

> The WebAuthn factor option is listed as **Security Key or Biometric**, as shown in the following screenshot.

Author: done

```

If the call is successful, the returned `enrollResponse` object has an `AuthenticationStatus` of `AwaitingAuthenticatorVerification` and its `CurrentAuthenticator` property contains the encrypted challenge and the user information that the browser can use to enroll the authenticator.

```csharp
Session["IdxContext"] = enrollResponse.IdxContext;
Session["isPasswordSelected"] = model.IsPasswordSelected;

switch (enrollResponse?.AuthenticationStatus)
{
    case AuthenticationStatus.AwaitingAuthenticatorVerification:
        {
            if (model.IsPasswordSelected)
            {
                return RedirectToAction("ChangePassword", "Manage");
            }
            else if (model.IsWebAuthnSelected)
            {
                Session["currentWebAuthnAuthenticator"] = 
                   enrollResponse.CurrentAuthenticator;
                return RedirectToAction(
                    "EnrollWebAuthnAuthenticator", "Manage");
            }

            return RedirectToAction("VerifyAuthenticator", "Manage");
        }

    // other case statements

    default:
        return View("SelectAuthenticator", model);
}
```

> ### 6: Enroll authenticator via the browser

Author: done

> Build a page which takes the challenge and user information from the website's backend servers and calls `navigator.credentials.create` to raise the prompt to enter a security key, validate with Windows Hello, Touch Id, or other WebAuthn Authenticator. For example, in the sample application, a new `EnrollWebAuthnViewModel` is populated from the `currentAuthenticator` object returned by the `enrollResponse` in the previous step.

Author: done

};
```

This `ViewModel` is then consumed in the script section of a Razor page. First it is prepared to be passed to the authenticator.

```js
document.addEventListener('DOMContentLoaded', function () {
    const challenge = '@Model.Challenge';
    const userId = '@Model.UserId';
    const username = '@Model.Username';
    const displayName = '@Model.DisplayName';

    const publicKeyCredentialCreationOptions = {
        rp: {
            name: "localhost",
            id: "localhost",
        },
        challenge: strToBin(challenge),
        user: {
            id: strToBin(userId),
            name: name,
            displayName: displayName,
        },
        pubKeyCredParams: [{alg: -7, type: "public-key"}],
    };
```

Then the call to `navigator.credentials.create` calls WebAuthn APIs in the browser and passes the challenge to the authenticator for validation. The user sees the authenticator challenge dialog box. In this case, the WebAuthn option is a security USB key.

![A dialog box generated by the browser prompting the user to set up their security key ](/img/authenticators/dotnet-Authenticators-WebAuthn-Enroll-KeySetup.png)

> When the authenticator has validated the user, it returns an `attestationObject` containing the information needed to validate the registration event, and a `clientDataJSON` object which is used to associate the new credential with the server and the browser.

Author: done

```js
    navigator.credentials.create({
        publicKey: publicKeyCredentialCreationOptions
    }).then((newCredential) => {
        const clientDataJSON = binToStr(
            newCredential.response.clientDataJSON);
        const attestationObject = binToStr(
            newCredential.response.attestationObject);

        const params = {
            "clientData": clientDataJSON,
            "attestation": attestationObject,
            "challenge": challenge,
            "userId": userId,
            "username": username,
            "displayName": displayName
        };

        const options = {
            method: 'POST',
            body: JSON.stringify(params),
            headers: { "Content-type": "application/json; charset=UTF-8" }
        };
```

### 7. Send verification credentials to Okta server

> Send the verifying credentials back to the Okta server to finish enrolling the WebAuthn authenticator for the user. Okta will decrypt the challenge using the user's public key and store the user's credentials and public key.

Author: done

```js
        fetch("@Url.Action("EnrollWebAuthnAuthenticatorAsync", "Manage")", options)
            .then(res => {
                console.log("Request successful! Response:", res);
                location.href = '@Url.Action("EnrollWebAuthnAuthenticator", "Manage",
                    new { verificationCompleted = true})';
            }).catch(function(err) {
                console.error(err);
                }
            );
    }).catch(function (err) {
        console.error(err);
    });
}, false);
```

Pass the credentials as a parameter to the `EnrollAuthenticatorAsync` method on the `IdxClient`.

```csharp
var authnResponse = await _idxClient.EnrollAuthenticatorAsync(
    new EnrollWebAuthnAuthenticatorOptions
    {
        Attestation = viewModel.Attestation,
        ClientData = viewModel.ClientData,
    }, (IIdxContext)Session["idxContext"]);

Session["WebAuthnResponse"] = authnResponse;
```

> ### 8: Enroll More Authenticators or Log the User In

Author: done

> Query the `AuthenticationStatus` property of the `AuthenticationResponse` object returned by `EnrollAuthenticatorAsync` to discover the current status of the authentication process. You'll need to respond to two specific authenticator statuses:

Author: done

* `AwaitingAuthenticatorEnrollment`
* `Success`

> A status of `AwaitingAuthenticatorEnrollment` means that there are other authenticator types (Google, Okta Verify) which can be enrolled by the user and that your app should return to Step 4 and display a list of those authenticators which are still unenrolled as well as an option to skip further enrollment.

Author: done

> A status of `Success` (or the user choosing to skip further authenticator enrollment) means that the user has now successfully enrolled their WebAuthn authenticator and is logged into the app. Call `AuthenticationHelper.GetIdentityFromTokenResponseAsync` to retrieve the OIDC claims information about the user and pass them into your application.

Reviewer: ```suggestion
A status of `Success` (or the user choosing to skip further authenticator enrollment) means that the user is now successfully enrolled using the WebAuthn authenticator and is signed in to the app. Call `AuthenticationHelper.GetIdentityFromTokenResponseAsync` to retrieve the OIDC claims information about the user and pass them into your application.
```

Author: @susanharper-okta  This change implies something slightly different from what I understand is happening.
